don't require address when adding/removing listener

trociny opened the pull request:
use the gateway address instead

Fixes: https://github.com/ceph/ceph-nvmeof/issues/13
Signed-off-by: Mykola Golub <mykola.golub@clyso.com>

trociny commented:
@sskaur @PepperJo

PepperJo reviewed (changes requested):
I think the listener configuration should be more flexible i.e. one should be able to have different number of listeners per gateway node and use different ports and possibly transports even if they are in the same group. What do you think?

trociny commented:
I don't have a strong opinion. I just thought that having the same listener configuration for all gateways would be enough and would simplify the configuration procedure for a user, and that it was the main idea of gateway groups: all gateways in a group behave the same. But it indeed it may be considered as a unnecessary limitation.

So I am also fine with your proposal, just would like to confirm we agree about the details:

I can imagine we would need to introduce "add(create) gateway" command, that would assign the gateway ID. Then in `create|delete_listener` commands, additionally to transport, address and port, we will need to specify the gateway ID, which will be used in the storage key, and a gateway would only use its own keys (listeners).

Is it what you had in your mind? Agree, it is much more flexible, though it will require more steps for a user, potentially executing almost the same command for every listener. And when a new gateway is added to a group, one will also need to create all listeners for this gateway.

I would also like to mention another use case, when gateways are running e.g. in k8s containers and the IP address is changed when a container is restarted. Not sure how useful this use case would be, I had it when preparing a demo, but in real life you probably would not want a gateway changing its IP after restart, so we might want to ignore this. Just mentioned for full picture.

PepperJo commented:
> I can imagine we would need to introduce "add(create) gateway" command, that would assign the gateway ID. Then in create|delete_listener commands, additionally to transport, address and port, we will need to specify the gateway ID, which will be used in the storage key, and a gateway would only use its own keys (listeners).

Yes, this is what I had in mind. The only headache is who is going to release IDs when the gateway is gone.

> Agree, it is much more flexible, though it will require more steps for a user, potentially executing almost the same command for every listener. And when a new gateway is added to a group, one will also need to create all listeners for this gateway.

Maybe we could allow to create listeners without gw id but just the port and use 0.0.0.0 to listen on. If the user is not interested in a more complex listener setup?

> I would also like to mention another use case, when gateways are running e.g. in k8s containers and the IP address is changed when a container is restarted. Not sure how useful this use case would be, I had it when preparing a demo, but in real life you probably would not want a gateway changing its IP after restart, so we might want to ignore this. Just mentioned for full picture.

This is not really solved with the local config solution since you would need to update it accordingly, no?
We can always leave the option to listen on 0.0.0.0

trociny commented:
> Maybe we could allow to create listeners without gw id but just the port and use 0.0.0.0 to listen on. If the user is not interested in a more complex listener setup?

LGTM. May be other people have different thoughts? One note though, for me (probably wrongly) "0.0.0.0" would mean listen all IPv4 addresses,  and "[::]" all IPv6. And I suppose we want either "any" or just "the gateway" ip that is specified in the static config. The second would be simpler (as it is currently implemented in this PR). To implement "any" would mean listing all node interfaces/addresses and creating listeners for every of this. Do we want something like this?

> This is not really solved with the local config solution since you would need to update it accordingly, no?

Yes, the config was auto generated on the container start.

PepperJo commented:
> LGTM. May be other people have different thoughts? One note though, for me (probably wrongly) "0.0.0.0" would mean listen all IPv4 addresses, and "[::]" all IPv6. And I suppose we want either "any" or just "the gateway" ip that is specified in the static config. The second would be simpler (as it is currently implemented in this PR). To implement "any" would mean listing all node interfaces/addresses and creating listeners for every of this. Do we want something like this?

Actually listening to `::` should be enough, on Linux machines it includes ipv4, so you don't need to go through all node interfaces etc. I have no strong opinion also allowing to set a static gateway ip in the local config. The only downside is that the user might be confused with all the options.

sdpeters commented:
> > LGTM. May be other people have different thoughts? One note though, for me (probably wrongly) "0.0.0.0" would mean listen all IPv4 addresses, and "[::]" all IPv6. And I suppose we want either "any" or just "the gateway" ip that is specified in the static config. The second would be simpler (as it is currently implemented in this PR). To implement "any" would mean listing all node interfaces/addresses and creating listeners for every of this. Do we want something like this?
> 
> Actually listening to `::` should be enough, on Linux machines it includes ipv4, so you don't need to go through all node interfaces etc. I have no strong opinion also allowing to set a static gateway ip in the local config. The only downside is that the user might be confused with all the options.

I think we need the ability to specify the IP address of the gateway, but I can see why some users would rather not have to specify it. If that argument is optional it can default to "any". 

It probably makes sense to take this in subnet form (a.b.c.d/s). That enables the user to put a gateway on a specific network, without knowing the gateway node's IP in that subnet. It also enables the gateway's exact IP to be specified (with /32, or by omitting /s). There could be a need to specify the exact IP when there are multiple gateways in the same node (as you might want in a two-socket dual-NIC OSD node when using ADNN), or if the user doesn't want the gateway to bind to all the gateway node's addresses in that subnet for some reason.

If this is agreeable, we can do it incrementally (e.g. make the IP parameter optional, and add subnet support later).

trociny commented:
Ok, so I am going to implement it in the following way:

- add `gateway_name` param in the static config. If not specified the hostname is used. It is going to be used to identify the gateway when creating/removing a listener (and storing in the persistent config).
- `create_listener` will continue to have address (`-a`) option, but optional now, and additionally it will have gateway (`-g`) option.

If both `-g` and `-a` option are specified the listener is created on this gateway with the specified address.

If only `-g` option is specified the listener is created on this gateway with the address from the gateway's config (`gateway_addr`).
 
If only `-a` option is specified the error "unknown gateway" is returned.

If no `-g` and `-a` options are specified the listener is created on all gateways with the address specified in the gateway's config.

Does it make sense?

PepperJo commented:
> Ok, so I am going to implement it in the following way:
> 
>     * add `gateway_name` param in the static config. If not specified the hostname is used. It is going to be used to identify the gateway when creating/removing a listener (and storing in the persistent config).
> 
>     * `create_listener` will continue to have address (`-a`) option, but optional now, and additionally it will have gateway (`-g`) option.
> 
> 
> If both `-g` and `-a` option are specified the listener is created on this gateway with the specified address.
> 
> If only `-g` option is specified the listener is created on this gateway with the address from the gateway's config (`gateway_addr`).
> 
> If only `-a` option is specified the error "unknown gateway" is returned.
> 
> If no `-g` and `-a` options are specified the listener is created on all gateways with the address specified in the gateway's config.
> 
> Does it make sense?

Makes sense to me. Maybe we can add one addition: if `gateway_addr` is not specified in the config listen on `::`.

trociny commented:
> Maybe we can add one addition: if gateway_addr is not specified in the config listen on `::`.

I am not sure what `::` would mean? What should I send in the `nvmf_subsystem_add_listener` command to the `nvmf_tgt` socket as the address argument? `::` does not work.

So for me `::` would mean "create listeners on all available addresses". I.e. the gw server would need to list all host's interfaces  and assigned addresses and call `nvmf_subsystem_add_listener` for every of them. I may do this. But the problem with this approach is that the interfaces and addresses may dynamically change (be added/removed) and the gatway would need to track this and update the `nvmf_tgt` accordingly.

Actually, my current implementation (see the code) for the case if gateway_addr is not specified (or `::` is specified) is to use the IP for a hostname returned by `gethostname`:
```
hostname = socket.gethostname()
self.traddr = socket.gethostbyname(hostname)
```

It might be not what users would want in all cases, but is simple and probably good enough?

PepperJo commented:
> I am not sure what `::` would mean? What should I send in the `nvmf_subsystem_add_listener` command to the `nvmf_tgt` socket as the address argument? `::` does not work.

Did you use addr family IPv6 otherwise you need to use 0.0.0.0

> So for me :: would mean "create listeners on all available addresses". I.e. the gw server would need to list all host's interfaces and assigned addresses and call nvmf_subsystem_add_listener for every of them. I may do this. But the problem with this approach is that the interfaces and addresses may dynamically change (be added/removed) and the gatway would need to track this and update the nvmf_tgt accordingly.

I don't think manually going through all interfaces is ever a good idea because of the mentioned issues.

> Actually, my current implementation (see the code) for the case if gateway_addr is not specified (or :: is specified) is to use the IP for a hostname returned by gethostname:

I'm not a big fan of this approach I had a fair share of headaches with this in the past.

trociny commented:
> Did you use addr family IPv6 otherwise you need to use 0.0.0.0

Indeed `nvmf_subsystem_add_listener  -a 0.0.0.0` seems to work, though not sure how well without through testing. The immediate issues found:

`nvme discover` returns `traddr:  0.0.0.0` (this actually may be not an issue, just does not look very useful, but what it could be?):
```
adonis:~/ceph/ceph-nvmeof.upstream% sudo nvme discover -t tcp -a 127.0.0.1 -s 4401                                        

Discovery Log Number of Records 1, Generation counter 6
=====Discovery Log Entry 0======
trtype:  tcp
adrfam:  ipv4
subtype: nvme subsystem
treq:    not required
portid:  0
trsvcid: 4401
subnqn:  nqn.2016-06.io.spdk:cnode1
traddr:  0.0.0.0
sectype: none
```
`nvme connect` fails:
```
adonis:~/ceph/ceph-nvmeof.upstream% sudo nvme connect -t tcp -n "nqn.2016-06.io.spdk:cnode1" -a 127.0.0.1 -s 4401
Failed to write to /dev/nvme-fabrics: Input/output error
```
In the tgt log:
```
[2022-05-25 12:38:06.765436] ctrlr.c: 680:nvmf_qpair_access_allowed: *ERROR*: Subsystem 'nqn.2016-06.io.spdk:cnode1' does not allow host 'nqn.2014-08.org.nvmexpress:uuid:00000000-0000-0000-0000-ac1f6b0c860a' to connect at this address.
```
Although I allowed any host:
```
adonis:~/ceph/ceph-nvmeof.upstream% sudo python3 ./spdk/scripts/rpc.py nvmf_subsystem_allow_any_host nqn.2016-06.io.spdk:cnode1 
adonis:~/ceph/ceph-nvmeof.upstream% sudo python3 ./spdk/scripts/rpc.py nvmf_get_subsystems                                      
[
  {
    "nqn": "nqn.2014-08.org.nvmexpress.discovery",
    "subtype": "Discovery",
    "listen_addresses": [],
    "allow_any_host": true,
    "hosts": []
  },
  {
    "nqn": "nqn.2016-06.io.spdk:cnode1",
    "subtype": "NVMe",
    "listen_addresses": [
      {
        "transport": "TCP",
        "trtype": "TCP",
        "adrfam": "IPv4",
        "traddr": "0.0.0.0",
        "trsvcid": "4401"
      }
    ],
    "allow_any_host": true,
    "hosts": [],
    "serial_number": "SPDK00000000000001",
    "model_number": "SPDK bdev Controller",
    "max_namespaces": 32,
    "min_cntlid": 1,
    "max_cntlid": 65519,
    "namespaces": [
      {
        "nsid": 1,
        "bdev_name": "test1",
        "name": "test1",
        "nguid": "65B171A5F192432BAE8FAFCCEA60920A",
        "uuid": "65b171a5-f192-432b-ae8f-afccea60920a"
      }
    ]
  }
]
```

PepperJo commented:
Thanks for testing it. This seems to be a bug in SPDK. 
Why don't we start with not allowing empty `gateway_addr`? And keep the rest as discussed. If others are in favor of using the hostname approach I'm also fine with it. @sskaur @idryomov ?

trociny committed:
improvements for adding/removing listener

Now you may create a gateway listener specifying the gateway name,
address and port, or you may create a group listener specifying
only the port. In the second case all gateways in the group will
create a listener using `gateway_addr` config param as `traddr`.

Fixes: https://github.com/ceph/ceph-nvmeof/issues/13

Signed-off-by: Mykola Golub <mykola.golub@clyso.com>

trociny commented:
Updated. I think it is ready for the next (or final) iteration of review.

PepperJo commented on nvme_gw_cli.py:
```diff
@@ -275,17 +275,26 @@ def delete_host(self, args):
     def create_listener(self, args):
         """Adds a listener at a particular TCP/IP address for a given subsystem."""
 
+        if args.gateway and not args.traddr:
```
Why not let the RPC fail with the error?

trociny replied:
I thought it would be more clear for users but I don't mind. I will change.

PepperJo commented on nvme_gw_cli.py:
```diff
@@ -298,17 +307,26 @@ def create_listener(self, args):
     def delete_listener(self, args):
         """Deletes a listener at a particular TCP/IP address for a given subsystem."""
 
+        if args.gateway and not args.traddr:
```
see above

PepperJo commented on nvme_gw_server.py:
```diff
@@ -62,6 +64,19 @@ def __init__(self, nvme_config):
         self.server = None
         self.transports = set()
 
+        self.gateway_name = self.nvme_config.get("config", "gateway_name")
```
We need to make sure `gateway_name` does not have `_` otherwise we probably have a problem with the persistent config. Generally, we might need to check that it only has characters that are valid in an OMAP key. We can let it fail later but than we run into the same semantics again that it fails late.

trociny replied:
Or use some other char instead of `_` as a separator. We have an issue for this: https://github.com/ceph/ceph-nvmeof/issues/10

PepperJo replied:
Ok yes, let's try to address this sooner than later. Otherwise it might be a mess to change. But I agree should be generic to all cases in another PR.

PepperJo commented on nvme_gw_server.py:
```diff
@@ -62,6 +64,19 @@ def __init__(self, nvme_config):
+        self.traddr = None
+        gateway_addr = self.nvme_config.get("config", "gateway_addr")
+        try:
+            ipaddr = ipaddress.ip_address(gateway_addr.strip('[]'))
```
Do we expect arrays in `gateway_addr`? We support one only anyway, no?

trociny replied:
Right now in the config example we ship we have [1]:
```
gateway_addr = [::]
```
I suppose it is not unusual practice to specify an ipv6 address in `[]` and `ipaddress` has problem with parsing it, so I had to strip it.

BTW, my current patch changes `gateway_addr = [::]` to `gateway_addr = 127.0.0.1`  in the config so the test that tests "add_listener" without gateway would work by default.

[1] https://github.com/ceph/ceph-nvmeof/blob/devel/nvme_gw.config#L13

PepperJo replied:
Ok it seems like the `[]` is used to distinguish between colons used after an address to specify the port in an URI [1], e.g. `[2001:db8:85a3:8d3:1319:8a2e:370:7348]:443`. I don't think we need to support `[]` syntax here.

[1] [https://datatracker.ietf.org/doc/html/rfc6874](https://datatracker.ietf.org/doc/html/rfc6874)

PepperJo commented on nvme_gw_server.py:
```diff
@@ -528,74 +545,102 @@ def nvmf_subsystem_remove_host(self, request, context=None):
+            try:
+                traddr = request.traddr
+                if not request.gateway:
+                    if request.trtype.lower() != "tcp":
```
Why do none tcp transports require a gateway?

trociny replied:
The idea is that in config's `gateway_addr` we specify an IP, and it makes for `tcp` transoport, but not for `user` or `pcie`. It looks like it also makes sense for `rdma` though I have no experience with RDMA to say if "gateway_addr" is going to be used in that case.

So my plan was for now (until we rely on `gateway_addr` config for it to work) to allow "add group listener" only for tcp transport. Do you think it would be too restrictive?

PepperJo replied:
I don't think `pcie` is allowed for NVMeoF. I see `fc`, `rdma`, `tcp` and `muser` in the code. At least for `rdma` it could be used in that way. For `fc` and `muser` we probably would need to allow to specify a `gateway_addr` that is not an ip. I don't have a strong opinion on it, on the other hand it would be nice if we could just support those out of the box. It seems like we would just need to remove trying to parse `gateway_addr` as an ip and let the create listener fail.

PepperJo commented on nvme_gw_server.py:
```diff
@@ -62,6 +64,19 @@ def __init__(self, nvme_config):
+            if not ipaddr.is_unspecified:
+                self.traddr = str(ipaddr)
+        except Exception as ex:
+            pass
```
Why don't we fail if the address is invalid?

trociny replied:
ack

trociny commented:
@PepperJo thanks. updated

PepperJo commented on nvme_gw_server.py:
```diff
@@ -528,74 +535,100 @@ def nvmf_subsystem_remove_host(self, request, context=None):
+                        raise Exception(f"{request.traddr} address reqires gateway")
+                    traddr = self.nvme_config.get("config", "gateway_addr")
+
+                if request.trtype.lower() not in self.transports:
```
We could just let SPDK handle this. It would avoid this duplication of code in add and delete listener. But I don't have a strong opinion on it. So feel free to leave it if you like it better this way.

trociny replied:
This check was added earlier, not in this PR. Personally I like a little more this way, because the error message seems a little more clear for users. But I don't mind letting SPDK handle this. I may add a commit that removes this `self.transports` thing.

trociny replied:
I pushed an additional commit for this for discussion/review

PepperJo reviewed (approved):
Thanks Mykola. Minor comment otherwise LGTM

trociny committed:
nvme_gw_server: don't check if transport is enabled

when adding/deleting a listener: let SPDK handle this.

Signed-off-by: Mykola Golub <mykola.golub@clyso.com>

PepperJo reviewed (approved):
LGTM

idryomov commented on nvme_gw_server.py:
```diff
@@ -528,74 +532,96 @@ def nvmf_subsystem_remove_host(self, request, context=None):
         """Adds a listener at the given TCP/IP address for the given subsystem."""
         self.logger.info({
-            f"Adding {request.trtype} listener at {request.traddr} : {request.trsvcid} for {request.nqn}"
+            f"Adding {request.gateway} {request.trtype} listener at {request.traddr} : {request.trsvcid} for {request.nqn}"
```
Get rid of spaces between the address and the port:
```suggestion
            f"Adding {request.gateway} {request.trtype} listener at {request.traddr}:{request.trsvcid} for {request.nqn}"
```

idryomov commented on nvme_gw_server.py:
```diff
@@ -528,74 +532,96 @@ def nvmf_subsystem_remove_host(self, request, context=None):
         """Removes a listener at the given TCP/IP address for the given subsystem."""
         self.logger.info(
-            {f"Removing {request.trtype} listener at {request.traddr} for {request.nqn}."})
+            {f"Removing {request.gateway} {request.trtype} listener at {request.traddr} : {request.trsvcid} for {request.nqn}"})
```
ditto

idryomov commented on nvme_gw_server.py:
```diff
@@ -60,7 +61,10 @@ def __init__(self, nvme_config):
         self.persistent_config = OmapPersistentConfig(nvme_config)
         self.spdk_process = None
         self.server = None
-        self.transports = set()
```
Drop `transports: List (set) of created transports` from the class comment.

idryomov commented on nvme_gw_server.py:
```diff
@@ -528,74 +532,96 @@ def nvmf_subsystem_remove_host(self, request, context=None):
+                traddr = request.traddr
+                if not request.gateway:
+                    if traddr:
+                        raise Exception(f"{request.traddr} address reqires gateway")
```
```suggestion
                        raise Exception(f"{request.traddr} address requires gateway")
```

idryomov replied:
Actually, this and the following "unknown traddr" check seem a bit hard to read to me.  I would suggest:
```
if not request.gateway and not request.traddr:
    traddr = self.nvme_config.get("config", "gateway_addr")
    if not traddr:
        raise Exception(f"config.gateway_addr option is not set")
else if request.gateway and request.traddr:
    traddr = request.traddr
else:
    raise Exception(f"both gateway and traddr must be specified")
```

idryomov commented on proto/nvme_gw.proto:
```diff
@@ -131,19 +131,21 @@ message req_status {
-	string adrfam = 4;
-	string traddr = 5;
-	string trsvcid = 6;
+	string gateway = 3;
```
I would name it `gateway_name` here and everywhere else in the code so that it is clear that it is the same thing as what the user specifies as `gateway_name` in the config file.

idryomov commented on nvme_gw_server.py:
```diff
@@ -528,74 +532,96 @@ def nvmf_subsystem_remove_host(self, request, context=None):
+                return pb2.req_status()
 
         if context:
             # Update persistent configuration
```
The persistent configuration is updated on gateway name mismatch with the expectation that all other gateways in the gateway group would see the change and attempt to execute the request, right?

If so, did you intend this update to be unconditional?  E.g. currently if a user specifies `gateway` but omits `traddr`, in case the handling gateway is the target gateway, the request is failed.  However, in case the handling gateway is not the target gateway, the persistent config is updated with an invalid record (which would be failed later by the target gateway, I assume).

trociny replied:
hope, it is better now

idryomov commented on nvme_gw_server.py:
```diff
@@ -528,74 +532,96 @@ def nvmf_subsystem_remove_host(self, request, context=None):
+                traddr = request.traddr
+                if not request.gateway:
+                    if traddr:
+                        raise Exception(f"{request.traddr} address reqires gateway")
```
ditto

trociny commented:
@idryomov thanks, updated

idryomov reviewed (changes requested):
LGTM but needs a rebase

trociny commented:
@idryomov thanks. rebased
